### PR TITLE
Collection of Various Improvements

### PR DESCRIPTION
**Pull request by Xaldew**
Hello!

I spent a bit of time over the last few months to improve this mode in various ways, most notably:

1. Make it a proper derivative of `cc-mode`, such that address qualifiers are properly handled.
2. Add support for almost all special OpenCL types and built-ins.
3. Renamed the primary file to prepare for a `opencl-c++-mode` if that is desired at some point in the future.
4. Update variable documentation where appropriate.

I'd figured I may as well contribute these back in case anyone else would like these features!

### CODE REVIEW

File: README.org [hunk 1]
```diff
@@ -1,45 +1,49 @@
-#+TITLE: Opencl-mode
-#+AUTHOR: BAH SALMANE
-#+EMAIL: salmane.bah@u-bordeaux.fr
-#+DATE: 25-06-2014
+#+TITLE: OpenCL Major Mode
+#+AUTHOR: Gustaf Waldemarson
```
Comment (salmanebah) [inline README.org:2]: Let's add all the authors

File: opencl-c-mode.el [hunk 2]
```diff
@@ -0,0 +1,808 @@
...
+  (interactive)
+  (let* ((api-function (if (region-active-p)
+                           (buffer-substring-no-properties (region-beginning)
+                                                           (region-end))
+                         (thing-at-point 'symbol)))
+         (doc-url (concat
+                   "http://www.khronos.org/registry/cl/sdk/2.1/docs/man/xhtml/"
+                   api-function ".html")))
+    (browse-url doc-url)))
+
```
Comment (salmanebah) [inline opencl-c-mode.el:806]: Let's also add  `(define-key opencl-mode-map (kbd "C-c ! d") 'opencl-lookup)`  for the mapping

File: README.org [hunk 3]
```diff
@@ -1,45 +1,49 @@
...
+#+EMAIL: gustaf.waldemarson ~at~ gmail.com
```
Comment (salmanebah) [inline README.org:3]: And email addresses

File: README.org [hunk 4]
```diff
@@ -1,45 +1,49 @@
...
+(use-package opencl-mode
+    :ensure t)
+#+end_src
+
+** Installing From Source
+
+1. Get the source code by cloning this repository:
    #+BEGIN_SRC sh
-   M-x package-install RET opencl-mode RET
+   $ git clone https://github.com/Xaldew/opencl-mode.git
```
Comment (salmanebah) [inline README.org:38]: We should also update this one to point to origin

**Review by salmanebah — COMMENTED** [review]
Thank you for the PR Gustaf !

File: README.org (hunk 3 above)
Comment (Xaldew) [reply README.org]: Hello!

Nice thing that you caught all the author changes, I had forgotten that I made those a while back when I was thinking I might have to take over ownership of the repo, which is clearly not necessary!

All things you have noted above have been fixed, however the: `(define-key opencl-mode-map (kbd "C-c ! d") 'opencl-lookup)` is now unnecessary, as it is added directly to the map when it is being defined on lines 713-717.

**Comment by Xaldew** [comment]
Hello!

Nice thing that you caught all the author changes, I had forgotten that I made those a while back when I was thinking I might have to take over ownership of the repo, which is clearly not necessary!

All things you have noted above have been fixed, however the: `(define-key opencl-mode-map (kbd "C-c ! d") 'opencl-lookup)` is now unnecessary, as it is added directly to the map when it is being defined on lines 713-717.

**Comment by salmanebah** [comment]
Thank you @Xaldew, PR approved !

**Review by salmanebah — APPROVED** [review]

**Pull request merged**